content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+
+Some examples of writing out joins in tidymodels packages: [tidymodels/parsnip#932](https://github.com/tidymodels/parsnip/pull/932), [tidymodels/parsnip#931](https://github.com/tidymodels/parsnip/pull/931), [tidymodels/parsnip#921](https://github.com/tidymodels/parsnip/pull/921), and [tidymodels/recipes#1121](https://github.com/tidymodels/recipes/pull/1121).
+
+### `nest()`
+
+`nest()`s are subject to similar considerations as joins. When they allow for expressive or principled user interfaces, use them, but manipulate them sparingly in backends. Writing out `nest()` calls _can_ result in substantial speedups, though, and the process is not quite as gnarly as writing out a join. For code that relies on `nest()`s and sees heavy traffic, rewriting with vctrs may be worth the effort.
+
+For example, consider nesting `mtcars_tbl` by `cyl`:
+
+```{r nest}
+nest(mtcars_tbl, .by = cyl)
+```
+
+For some basic nests, like nesting by values in one column, `vec_split()` can do the trick.
+
+```{r nest-alt}
+res <- 
+  vec_split(
+    x = mtcars_tbl[!colnames(mtcars_tbl) == "cyl"],
+    by = mtcars_tbl$cyl
+  )
+
+new_tibble(list(cyl = res$key, data = res$val))
+```
+
+The performance improvement in these situations can be quite substantial:
+
+```{r bm-nest}
+bench::mark(
+  nest = nest(mtcars_tbl, .by = cyl),
+  vctrs = {
+    res <-
+      vec_split(
+        x = mtcars_tbl[!colnames(mtcars_tbl) == "cyl"],
+        by = mtcars_tbl$cyl
+      )
+    new_tibble(list(cyl = res$key, data = res$val))
+  }
+) %>%
+  select(expression, median)
+```
+
+More complex nests require a good bit of facility with the vctrs package. `vec_split()`, `list_unchop()`, and `vec_chop()` are all good places to start, and these examples of writing out nests in tidymodels packages make use of other vctrs patterns: [tidymodels/tune#657](https://github.com/tidymodels/tune/pull/657), [tidymodels/tune#657](https://github.com/tidymodels/tune/pull/656), [tidymodels/tune#640](https://github.com/tidymodels/tune/pull/640), and [tidymodels/recipes#1121](https://github.com/tidymodels/recipes/pull/1121).
+
+### Combining strings
+
+The glue package is super helpful for writing expressive and correct strings with data. It is quite a bit slower than `paste0()`, though.
+
+```{r bm-glue}
+name <- "Simon"
+
+bench::mark(
+  glue = glue::glue("My name is {name}."),
+  paste0 = paste0("My name is ", name, "."),
+  check = FALSE
+) %>% 
+  select(expression, median)
+```
+
+My rule of thumb is to use `glue()` for errors, when the function will stop executing anyway. For simple pastes that are intended to be called repeatedly, use `paste0()`. There's a lot of gray area in between those two contexts---intuit (or profile) as you will.
```

Comment:
You should warn about `paste0("prefix-", character())` not following tidyverse recycling rules, possibly resulting in lots of bugs. A decent / safer middle ground is:

```r
vec_paste0 <- function (...) {
  args <- vec_recycle_common(...)
  exec(paste0, !!!args)
}
```

Reply:
And if you can require a new enough R version, you can use `paste0(recycle0 = TRUE)`